Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch

```diff
@@ -0,0 +1,294 @@
+index 03c15e7c9..1f987d6e6 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+@@ -8,6 +8,7 @@ import java.io.IOException;
+ import java.lang.reflect.InvocationTargetException;
+ import java.lang.reflect.Method;
+ import java.lang.reflect.Modifier;
++import java.nio.charset.StandardCharsets;
+ import java.util.HashMap;
+ import java.util.List;
+ import java.util.Map;
+@@ -69,7 +70,7 @@ public class PaperConfig {
+         commands.put("paper", new PaperCommand("paper"));
+ 
+         version = getInt("config-version", 14);
+-        set("config-version", 15);
++        set("config-version", 16);
+         readConfig(PaperConfig.class, null);
+     }
+ 
+@@ -248,7 +249,7 @@ public class PaperConfig {
+     private static void bungeeOnlineMode() {
+         bungeeOnlineMode = getBoolean("settings.bungee-online-mode", true);
+     }
+-    
++
+     public static int packetInSpamThreshold = 300;
+     private static void packetInSpamThreshold() {
+         if (version < 11) {
+@@ -433,4 +434,26 @@ public class PaperConfig {
+             }
+         }
+     }
++
++    public static boolean velocitySupport;
++    public static boolean velocityOnlineMode;
++    public static byte[] velocitySecretKey;
++    private static void velocitySupport() {
++        if (version < 16) {
++            ConfigurationSection section = config.createSection("settings.velocity-support");
++            section.set("enabled", false);
++            section.set("online-mode", true);
++            section.set("secret", "");
++        }
++
++        velocitySupport = getBoolean("settings.velocity-support.enabled", false);
++        velocityOnlineMode = getBoolean("settings.velocity-support.online-mode", false);
++        String secret = getString("settings.velocity-support.secret", "");
++        if (velocitySupport && secret.isEmpty()) {
++            logError("Velocity support is enabled, but no secret key was specified. Velocity support will not be enabled.");
++            velocitySupport = false;
++        } else {
++            velocitySecretKey = secret.getBytes(StandardCharsets.UTF_8);
++        }
++    }
+ }
+diff --git a/src/main/java/com/destroystokyo/paper/profile/CraftPlayerProfile.java b/src/main/java/com/destroystokyo/paper/profile/CraftPlayerProfile.java
+index b151a13c1..2946722bb 100644
+--- a/src/main/java/com/destroystokyo/paper/profile/CraftPlayerProfile.java
++++ b/src/main/java/com/destroystokyo/paper/profile/CraftPlayerProfile.java
+@@ -146,7 +146,8 @@ public class CraftPlayerProfile implements PlayerProfile {
+         UserCache userCache = server.getUserCache();
+         if (profile.getId() == null) {
+             final GameProfile profile;
+-            boolean isOnlineMode = server.getOnlineMode() || (SpigotConfig.bungee && PaperConfig.bungeeOnlineMode);
```

Comment:
We really need to abstract all this now to a single method PaperConfig.isProxyOnlineMode()